### PR TITLE
Topic/incr build improvs pr

### PR DESCRIPTION
**Pull request by danielocfb**
Assortment of improvements to incremental build logic. Please see individual commits for descriptions.

### CODE REVIEW

**Comment by mykola-lysenko** [comment]
nit: "but with ISO 8601 Git would simply encode the
committer's time zone can directly be encoded" - this one reads strange, in the first commit message

**Comment by danielocfb** [comment]
> nit: "but with ISO 8601 Git would simply encode the committer's time zone can directly be encoded" - this one reads strange, in the first commit message

Thanks, I fixed it up.

**Review by mykola-lysenko — APPROVED** [review]

**Comment by mykola-lysenko** [comment]
Let's wait for tests to pass before merging

File: .github/workflows/test.yml
Comment (chantra) [inline .github/workflows/test.yml:194]: :exploding_head:

**Review by chantra — APPROVED** [review]

File: .github/workflows/test.yml
Comment (chantra) [inline .github/workflows/test.yml:149]: I wonder if we could get kpd to add a tag to the commits it generates, this way we could strip the topmost commits with the tag present.
Reasoning is that the day we add a new commit via KPD, this will break.

**Review by chantra — APPROVED** [review]

File: .github/workflows/test.yml
Comment (danielocfb) [reply .github/workflows/test.yml:149]: The problem I see is that then we rely on the tag being present itself and the breakage of it not being there would be more severe (`HEAD^` is always present, but it may reference a non-upstream commit or one earlier than expected; neither conditions really break anything here, as I tried to explain in the summary). We can guard against the tag not being there, but then we are back at square one on the fallback path.

Ideally we'd need neither and have a way to identify something as upstream. We could do tricks such as take the first commit with a `Link:` tag in its summary, but I felt that's not more stable either. My preference would be to use something like `git-merge-base` against a known branch, but I haven't been able to make that work -- the clone/checkout we work on didn't have a whole lot of stuff to work with. I can play around some more and see if I can improve it.

Comment (danielocfb) [reply .github/workflows/test.yml:149]: Going to land the change first, though, as I still think it's an improvement.

**Pull request merged**
